src/main/java/gyro/google/storage/BucketResource.java

```diff
@@ -0,0 +1,468 @@
+            throw new GyroException(e.getDetails().getMessage());
+        }
+    }
+
+    @Override
+    public void update(GyroUI ui, State state, Resource current, Set<String> changedFieldNames) throws Exception {
+        Storage storage = creatClient(Storage.class);
+
+        try {
+            Bucket bucket = storage.buckets().get(getName()).execute();
+            bucket.setAcl(getAcl() == null ? null : getAcl().stream().map(BucketAccessControlConfiguration::toBucketAccessControl).collect(Collectors.toList()));
+            bucket.setLabels(getLabels());
+            bucket.setLocation(getLocation());
+            bucket.setDefaultEventBasedHold(getDefaultEventBasedHold());
+            bucket.setCors(getCors().stream().map(BucketCors::toBucketCors).collect(Collectors.toList()));
+            bucket.setBilling(getBilling() == null ? null : getBilling().toBucketBilling());
+            bucket.setEncryption(getEncryption() == null ? null : getEncryption().toBucketEncryption());
+            bucket.setIamConfiguration(getIamConfiguration() == null ? null : getIamConfiguration().toBucketIamConfiguration());
+            bucket.setLifecycle(getLifecycle() == null ? null : getLifecycle().toLifecycle());
+            bucket.setLogging(getLogging() == null ? null : getLogging().toBucketLogging());
+            bucket.setRetentionPolicy(getRetentionPolicy() == null ? null : getRetentionPolicy().toBucketRententionPolicy());
+            bucket.setStorageClass(getStorageClass());
+            bucket.setVersioning(getVersioning() == null ? null : getVersioning().toBucketVersioning());
+            bucket.setWebsite(getWebsite() == null ? null : getWebsite().toBucketWebsite());
+
+            // Lock retention policy. This can not be undone and ALL assets must reach policy time to delete bucket.
+            if (changedFieldNames.contains("retention-policy")
+                    && Boolean.TRUE.equals(bucket.getRetentionPolicy().getIsLocked())) {
+                storage.buckets().lockRetentionPolicy(getName(), bucket.getMetageneration()).execute();
+            }
+
+            storage.buckets().update(getName(), bucket).execute();
+        } catch (GoogleJsonResponseException e) {
+            throw new GyroException(e.getDetails().getMessage());
+        }
+    }
+
+    @Override
+    public void delete(GyroUI ui, State state) throws Exception {
+        try {
+            Storage storage = creatClient(Storage.class);
+            storage.buckets().delete(getName()).execute();
+        } catch (IOException e) {
+            throw new GyroException(String.format("Unable to delete Bucket: %s, Google error: %s", getName(), e.getMessage()));
+        }
+    }
+
+    @Override
+    public List<ValidationError> validate() {
+        List<ValidationError> errors = new ArrayList<>();
+
+        if (getName() != null && NAME_PATTERN.matcher(getName()).find()) {
+            errors.add(new ValidationError(this, "name", "Invalid name format."));
```

Comment:
If we do not handle validating this, does the api throw a meaningful error message? Currently `Invalid name format.` does not give the user much info on what the correct format is. Either we can let the api handle it if the message is clear and descriptive, or we should include details of what the correct format of the name should be in the message.

Reply:
I'm going to let Google handle the error.